main/tests/test_oa_hash_map.cpp

```diff
@@ -210,6 +210,65 @@ MainLoop *test() {
 		}
 	}
 
+	// Test map with 0 capacity.
+	{
+		OAHashMap<int, String> original_map(0);
+		original_map.set(1, "1");
+		OS::get_singleton()->print("OAHashMap 0 capacity test passed.\n");
+	}
+
+	// Test copy constructor.
```

Comment:
These comments are both about the copy constructor and assignment operator tests;
- `CRASH_COND` is a bit too aggresive. I think it's better to do the same as in the other cases: check the post-conditions by code and print an informative error message if they fail.
- Another interesting post-condition to check is the capacity of the copy. But currently there's no easy way to check it. We don't have yet a formal testing procedure about how test cases can delve into the private side or classes, etc. I think we can get by this time by just adding a public `get_capacity()`.

Reply:
External unit tests like this should just check the interface correctness and not internal implementation.

I think it's fine don't check the **real** capacity of the map, since the allocated capacity is an implementation detail.

Regarding `CRASH_COND` it should be the way how the unit test work. It's a way to alert the CI that the test didn't pass, not the other way around to notify the user with a text message.

Reply:
About`CRASH_COND`:
Unless the code crashes inevitably, or, in other words, as long as you can check the post-conditions in a controlled manner, it's better to do so. That way you can give some informative message about why it failed.
But let's assume I'm wrong. Then, since we already have a mix of "testing frameworks" in different test classes, 
in order to avoid introducing yet another approach, it would be better to modify all the cases across all of them to behave the same.
That's especially true given that changing the convention of how a test reports failure from what they do now to crashing is a fundamental change that should be propagated.
Since that's maybe too impractical at this point, I believe the best compromise now is at least to make new tests in this class follow the same principles as the others already present.

About capacity:
Sometimes you have no other option that doing something special to the class being tested to allow the test to check interesting post-conditions.
I agree that these tests are written under a black box perspective and therefore, since the capacity is private, they shouldn't just check it.
My point is that, until we are able to settle in a convention about how to look under the hood (helper methods for testing, tests are friends of the tested classes or whatever), a good balance is reached by adding a getter for capacity so the tests don't need to violate the strict black box principle and at the same time they can check something as important as that.
In other cases it may be unjustified and there would be a temptation to grow the public API to artificially expose internals that should remain private.
In this case I think it's not that bad (a getter for capacity is quite standard) and the benefit is huge.
For instance, without doing the capacity getter, the test case in the other comment thread run the risk of not testing anything. It's relying on the fact that a capacity of zero will make it crash, but there's no guarantee that such a thing will happen.

Reply:
I see what you mean regarding the `CRASH_COND`, and yes I agree that is better to not introduce another convention that is not fully used anyway. So I'm going to remove it.

___

Regarding the `capacity`, I'm not against making it public (even because it's already public: https://github.com/godotengine/godot/blob/master/core/oa_hash_map.h#L181). I don't like too much the idea of checking it here, because it's an internal implementation detail and this test should asses only the correctness of the public API.

For example, why would you check the capacity of the `Vector` here? If later you improve the algorithm to handle the `Vector` capacity, this test should help to assess that the exposed API still valid.
This test mimics the user usage, and the user don't need to know how the capacity is handled internally.

So, I think that this specific test should only test the API; if we want to test internal stuff, we need to add such feature to the _Unit test fremework_ so to be able to code the test directly into the file or the class.

Reply:
`CRASH_COND` removed.

Reply:
I'm still not entirely convinced, but I don't think this is super important now either.